docs/contribute-add-feature.md

```diff
@@ -0,0 +1,460 @@
+
+Before moving on to the next step, let's change to the `mystmd` directory
+```shell
+$ cd mystmd
+```
+From this point in the guide, terminal sessions will show the current working directory before the `$` prefix, excluding the path to the `mystmd` directory itself e.g.
+```shell
+$ echo MyST is cool!
+MyST is cool!
+$ cd packages
+(packages)$ echo MyST is cool!
+MyST is cool!
+```
+
+(defining-roles)=
+## Defining a Role
+The core specification for MyST as a markup language is defined in [the MyST spec](https://mystmd.org/spec). Most features in MyST should, over time, be incorporated into this specification so that consumers of MyST documents (such as `myst-parser` from the Jupyter Book software stack) can agree on the manner in which the contents should be parsed and rendered. Despite its importance, we can ignore exploring the process of updating the specification for this guide.
+
+We should begin by asking the question "What is a role?" The spec [defines roles](https://mystmd.org/spec/overview#roles) as 
+> similar to directives, but they are written entirely in one line.
+
+We want to create a _new_ `word-count` role that injects the total word count into a document. It should accept a format-string that allows us to format the resulting text, i.e.
+```markdown
+This is a lengthy document ...
+
+{word-count}`The number of words in this document is {number} words`
+```
+should become
+```markdown
+This is a lengthy document ...
+
+The number of words in this document is 5 words
+```
+
+Many of the "core" roles in `mystmd` are implemented in the [`myst-roles` package](https://github.com/executablebooks/mystmd/tree/main/packages/myst-roles). Although a word-count role might not be considered a "core" feature, we will pretend it is for this tutorial. Let's start by looking at the existing `abbreviation` role in [`packages/myst-roles/src/abbreviation.ts`](https://github.com/executablebooks/mystmd/blob/main/packages/myst-roles/src/abbreviation.ts)
+
+:::{tip}
+You can hover your mouse cursor over the link to [`packages/myst-roles/src/abbreviation.ts`](https://github.com/executablebooks/mystmd/blob/main/packages/myst-roles/src/abbreviation.ts) to see the contents of the file.
+:::
+
+We can see that `abbrevationRole` is annotated with the type `RoleSpec`. This is the basic type of a role declaration defined by the MyST specification. There are a number of important fields, such as the `name`, `alias`, and `body`. Our role will have the name `word-count`, and knowing that, we can define a barebones implementation that doesn't do anything! Let's add a new source file[^src] `word-count.ts` in the `myst-roles` package, and write the following:
+```{code-block} typescript
+:filename: packages/myst-roles/src/word-count.ts
+
+import type { RoleSpec, RoleData, GenericNode } from 'myst-common';
+
+export const wordCountRole: RoleSpec = {
+  name: 'word-count',
+  body: {
+    type: String,
+    required: true,
+  },
```

Comment:
These are default available fields right? Can you add an example with adding a new role option so that the API is a bit clearer?

Reply:
Oh, I guess I'm confusing `directive` with `role`? I guess there are some `directive` that wrap the call of `role` in those cases :thinking:. For another documentation than.